README
This is Perl module Plack::Middleware::AxsLog.

INSTALLATION

Plack::Middleware::AxsLog installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Plack::Middleware::AxsLog

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Plack::Middleware::AxsLog documentation is available as in POD. So you can do:

    % perldoc Plack::Middleware::AxsLog

to read the documentation online with your favorite pager.

Masahiro Nagano


Changes
Revision history for Perl extension Plack::Middleware::AxsLog

{{$NEXT}}

0.21 2014-02-20T06:14:17Z

        - [BUGFIX] log_line that returns from ALFC coderef does not have "\n"

0.20 2014-01-29T03:02:52Z

        - use POSIX::strftime::Compiler 0.30

0.13 2013-10-27T14:15:09Z

        - add format_options (Thank you zentooo)

0.12 2013-09-02T06:06:34Z

        - re-packaging

0.11 2013-04-09T15:44:33Z

        - updated POD
        - re-packaging

0.10    Sun Mar 3 16:38:54 2012
        - supports any format with Apache::LogFormat::Compiler
          be 3% faster than previous version

0.03    Fri Feb 8 13:56:20 2013
        - supports LTSV format

0.02    Tue Oct 9 11:45:11 2012
        - updated doc

0.01    Wed Aug 22 16:11:51 2012
        - original version


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Plack-Middleware-AxsLog',
    module_name     => 'Plack::Middleware::AxsLog',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Plack/Middleware/AxsLog.pm
package Plack::Middleware::AxsLog;

use strict;
use warnings;
use 5.8.5;
use parent qw/Plack::Middleware/;
use Plack::Util;
use Time::HiRes qw/gettimeofday/;
use Plack::Util::Accessor qw/response_time combined ltsv format format_options compiled_format error_only long_response_time logger/;
use POSIX qw//;
use Time::Local qw//;
use HTTP::Status qw//;
use Apache::LogFormat::Compiler;

our $VERSION = '0.20';

sub prepare_app {
    my $self = shift;
    $self->combined(1) if ! defined $self->combined;
    $self->response_time(0) if ! defined $self->response_time;
    $self->error_only(0) if ! defined $self->error_only;
    $self->long_response_time(0) if ! defined $self->long_response_time;

    my ($format, %format_options);
    if ( $self->format ) {
        $format = $self->format;
        %format_options = %{ $self->format_options } if $self->format_options;
    }
    elsif ( $self->ltsv ) {
        $format = join "\t",
            qw!host:%h user:%u time:%t req:%r status:%>s size:%b referer:%{Referer}i ua:%{User-agent}i!;
        $format .= "\t" . 'taken:%D' if $self->response_time;
    }
    elsif ( $self->combined ) {
        $format = q!%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i"!;
        $format .= ' %D' if $self->response_time;
    }
    else {
        $format = q!%h %l %u %t "%r" %>s %b!;
        $format .= ' %D' if $self->response_time;
    }

    $self->compiled_format(Apache::LogFormat::Compiler->new($format, %format_options)->code_ref);
}

sub call {
    my $self = shift;
    my($env) = @_;

    my $t0 = [gettimeofday];

    my $res = $self->app->($env);
    if ( ref($res) && ref($res) eq 'ARRAY' ) {
        my $length = Plack::Util::content_length($res->[2]);
        if ( defined $length ) {
            $self->log_line($t0, $env,$res,$length);
            return $res;
        }        
    }
    Plack::Util::response_cb($res, sub {
        my $res = shift;
        my $length = Plack::Util::content_length($res->[2]);
        if ( defined $length ) {
            $self->log_line($t0, $env,$res,$length);
            return;
        }
        return sub {
            my $chunk = shift;
            if ( ! defined $chunk ) {
                $self->log_line($t0, $env,$res,$length);
                return;
            }
            $length += length($chunk);
            return $chunk;
        };	
    });
}

sub log_line {
    my $self = shift;
    my ($t0, $env, $res, $length) = @_;

    my $elapsed = int(Time::HiRes::tv_interval($t0) * 1_000_000);

    unless (
         ( $self->{long_response_time} == 0 && !$self->{error_only} )
      || ( $self->{long_response_time} != 0 && $elapsed >= $self->{long_response_time} ) 
      || ( $self->{error_only} && HTTP::Status::is_error($res->[0]) ) 
    ) {
        return;
    }
    my $log_line = $self->{compiled_format}->(
        $env,
        $res,
        $length,
        $elapsed,
        $t0->[0],
    );

    if ( ! $self->{logger} ) {
        $env->{'psgi.errors'}->print($log_line."\n");
    }
    else {
        $self->{logger}->($log_line."\n");
    }
}


1;
__END__

=head1 NAME

Plack::Middleware::AxsLog - Yet another AccessLog Middleware

=head1 SYNOPSIS

  use Plack::Builder;

  builder {
      enable 'AxsLog',
        combined => 1,
        response_time => 1,
        error_only => 1,
      $app
  };

=head1 DESCRIPTION

Alternative implementation of Plack::Middleware::AccessLog. 
This middleware supports response_time and content_length calculation 
AxsLog also can set condition to display logs by response_time and status code.

Originally, AxsLog was faster AccessLog implementation. But PM::AccessLog became 
to using same access-log generator module L<Apache::LogFormat::Compiler>. 
Two middlewares have almost same performance now.

=head1 ARGUMENTS

=over 4

=item combined: Bool

log format. if disabled, "common" format used. default: 1 (combined format used)

common (Common Log Format) format is

  %h %l %u %t \"%r\" %>s %b
  
  => 127.0.0.1 - - [23/Aug/2012:00:52:15 +0900] "GET / HTTP/1.0" 200 645

combined (NCSA extended/combined log format) format is

  %h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-agent}i\"
  
  => 127.0.0.1 - - [23/Aug/2012:00:52:15 +0900] "GET / HTTP/1.1" 200 645 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_8) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.79 Safari/537.1"


=item ltsv: Bool

use ltsv log format. default: 0

LTSV (Labeled Tab-separated Values) format is

  host:%h<TAB>user:%u<TAB>time:%t<TAB>req:%r<TAB>status:%>s<TAB>size:%b<TAB>referer:%{Referer}i<TAB>ua:%{User-agent}i
  
  => host:127.0.0.1<TAB>user:-<TAB>time:[23/Aug/2012:00:52:15 +0900]<TAB>req:GET / HTTP/1.1<TAB>status:200<TAB>size:645<TAB>"referer:-<TAB>ua:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_6_8) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.79 Safari/537.1

See also L<http://ltsv.org/>

=item format: String

A format string.

  builder {
      enable 'AxsLog', 
          format => '%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i" %D';
      $app
  };

See details on perldoc L<Apache::LogFormat::Compiler>

=item format_options

This variable is passed to L<Apache::LogFormat::Compiler>. You can add char_handlers
and block_handlers with this middleware.

    enable 'AxsLog', 
        format => '%z %{X_MYAPP_VARIABLE}Z', 
        format_options => +{
            char_handlers => +{
                'z' => sub { 'z' },
            },
            block_handlers => +{
                'Z' => sub { 'Z' },
            },
        };

=item response_time: Bool

Adds time taken to serve the request. default: 0. This args effect to common, combined and ltsv format.

=item error_only: Bool

Display logs if response status is error (4xx or 5xx). default: 0

=item long_response_time: Int (microseconds)

Display log if time taken to serve the request is above long_response_time. default: 0 (all request logged)

=item logger: Coderef

Callback to print logs. default:none ( output to psgi.errors )

  use File::RotateLogs;
  my $logger = File::RotateLogs->new();

  builder {
      enable 'AxsLog',
        logger => sub { $logger->print(@_) }
      $app
  };

=back

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo {at} gmail.comE<gt>

=head1 SEE ALSO

L<Plack::Middleware::AccessLog>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


logbench.pl
#!/usr/bin/env perl

use strict;
use warnings;
use HTTP::Request::Common;
use HTTP::Message::PSGI;
use Plack::Test;
use Plack::Builder;
use Benchmark qw/cmpthese timethese/;

my $app = builder {
    sub{ [ 200, [], [ "Hello "] ] };
};

my $log_app = builder {
    enable 'AccessLog', format => "combined";
    sub{ [ 200, [], [ "Hello "] ] };
};


my $axslog_app = builder {
    enable 'AxsLog', combined => 1, response_time => 1;
    sub{ [ 200, [], [ "Hello "] ] };
};

my $axslog_format_app = builder {
    enable 'AxsLog', format => '%h %l %u %t "%r" %>s %b "%{Referer}i" "%{User-agent}i" %D';
    sub{ [ 200, [], [ "Hello "] ] };
};


my $axslog_error_only_app = builder {
    enable 'AxsLog', combined => 1, response_time => 1, error_only => 1;
    sub{ [ 200, [], [ "Hello "] ] };
};


my $env = req_to_psgi(GET "/");
open(STDERR,'>','/dev/null');

cmpthese(timethese(0,{
    'nolog' => sub {
        $app->($env);
    },
    'log'   => sub {
        $log_app->($env);
    },
    'axslog'   => sub {
        $axslog_app->($env);
    },
    'axslog_format'   => sub {
        $axslog_format_app->($env);
    },
    'error_only_axslog'   => sub {
        $axslog_error_only_app->($env);
    }
}));

__END__
Benchmark: running axslog, axslog_format, error_only_axslog, log, nolog for at least 3 CPU seconds...
    axslog:  4 wallclock secs ( 3.17 usr +  0.02 sys =  3.19 CPU) @ 16083.39/s (n=51306)
axslog_format:  3 wallclock secs ( 3.02 usr +  0.02 sys =  3.04 CPU) @ 15986.84/s (n=48600)
error_only_axslog:  4 wallclock secs ( 3.20 usr +  0.03 sys =  3.23 CPU) @ 41389.16/s (n=133687)
       log:  3 wallclock secs ( 3.05 usr +  0.02 sys =  3.07 CPU) @ 3200.65/s (n=9826)
     nolog:  3 wallclock secs ( 3.18 usr +  0.01 sys =  3.19 CPU) @ 438384.64/s (n=1398447)
                      Rate    log axslog_format axslog error_only_axslog   nolog
log                 3201/s     --          -80%   -80%              -92%    -99%
axslog_format      15987/s   399%            --    -1%              -61%    -96%
axslog             16083/s   403%            1%     --              -61%    -96%
error_only_axslog  41389/s  1193%          159%   157%                --    -91%
nolog             438385/s 13597%         2642%  2626%              959%      --


